### PR TITLE
configure : Add PowerPC64 building

### PR DESCRIPTION
**Pull request by jfoug**
Adds logic required to build PowerPC64 into the configure script
Should add altivec SIMD also.

### CODE REVIEW

**Comment by jfoug** [comment]
NOTE, on my QEMU vm, I have to configure like this:

```
$  CFLAGS=-I/usr/local/ppc64/include LDFLAGS=-L/usr/local/ppc64/lib ./confiugre
```

This is because I had to build oSSL in 64 bit mode (and also did GMP).  I used the local path /usr/local/ppc64 when building/installing them.   I could not find 64 bit libraries for most things, so installed thusly.  I also did not want to pollute local locations like /lib64  /usr/include (etc), since there were 32 bit items there installed using apt-get.  I figured since these 64 bit libs were not part of the apt-get universe, that I would only shoot myself in the foot installing into system dirs.

If someone has any better insite on how to properly get this to a better 64 bit build env, please let me know.

I am wondering if setting CFLAGS and LDFLAGS extra stuff into my ~/.bashrc might be a good thing? (also LD_LIBRARY_PATH)  I am no unix master, so help would really be appreciated.

**Comment by jfoug** [comment]
Note, one thing to review, is this patch moves the testing for -m64 (i.e. 64 bit build) to much earlier in the process.  This is due to -m64 needing to be added before ANY checking for type sizes. i.e. the 32 bit compiler was being used for type testing, so things like LONG were being sized at 4 bytes.   

I really think moving the 64 bit testing, and making sure than any 64 bit flags needed are added very early in the configure process IS a much better place.    I have seen no bad side effects of moving that logic to an early stage, tested on cygwin, linux  (64 and 32 bits), along with the PPC64 coding.  The PPC needed it, (as do other systems), but I did not want to add exceptional logic for just a couple systems.

I bet now with this change, that the sparc boxes I have at work, would also build (or 'could' build) in 64 bit mode  (UNTESTED).  They are 64 bit iron, but the 'defaults' are set to build 32 bit stuff (just like I was seeing on the Debian PPC64 install I did).  

But I would like others opinions on this, and if there ARE latent problems, now is the time to address this.

**Comment by kholia** [comment]
> I am wondering if setting CFLAGS and LDFLAGS extra stuff into my ~/.bashrc might be a good thing? (also LD_LIBRARY_PATH)

This doesn't sound like a good idea on regularly used systems.

**Comment by magnumripper** [comment]
>> I am wondering if setting CFLAGS and LDFLAGS extra stuff into my ~/.bashrc might be a good thing? (also LD_LIBRARY_PATH)
>
> This doesn't sound like a good idea on regularly used systems.

Well if it's the .bashrc local to that QEMU thingy, I can't see any problem with it. Personally I'm a big fan of bash history though. I set history size to something ridiculously high and never note anything down anywhere else 😆

**Comment by magnumripper** [comment]
On the bash history subject, here's what I do on ALL machines I have to deal with
```
$ cat ~/.inputrc 
"\e[6~": history-search-forward
"\e[5~": history-search-backward
```
After that (logout and login again) I can just type `./con` and hit page-up until I see whatever `./configure` line I want from the history.

**Comment by jfoug** [comment]
>   hit page-up until I see whatever ./configure line I want from the history.

LOL, F'n love that tip.

**Comment by jfoug** [comment]
Btw, the way things have been building (size testing prior to checking and forcing 64 bit), was not a 'good thing' (tm)   I know I beat my head over the sparc system at work, and finally gave up on trying to get 64 bit working, without doing lots of hand editing.   But I bet I could do it now, given time.

**Comment by solardiz** [comment]
I just did on `gcc110`:

```
git clone https://github.com/jfoug/JohnTheRipper -b ppc64 jfoug-ppc64
cd jfoug-ppc64/src/
./configure 
```

got this:

```
Target CPU ................................. powerpc64 ALTIVEC, 64-bit BE
AES-NI support ............................. no
Target OS .................................. linux-gnu
Cross compiling ............................ no
Legacy arch header ......................... ppc64.h
```

and it built cleanly:

```
$ make -s clean && make -sj64
ar: creating aes.a
stribog_fmt_plug.c:495:2: warning: #warning Stribog-256 and Stribog-512 formats require SSE 4.1, formats disabled [-Wcpp]
 #warning Stribog-256 and Stribog-512 formats require SSE 4.1, formats disabled
  ^
ar: creating secp256k1.a

Make process completed.
```

but then somehow it fails with:

```
$ ../run/john -test
Will run 64 OpenMP threads
Benchmarking: descrypt, traditional crypt(3) [DES 128/128 AltiVec]... (64xOMP) Illegal instruction
```

although some other formats work:

```
$ ../run/john -test -form=md5crypt
Will run 64 OpenMP threads
Benchmarking: md5crypt, crypt(3) $1$ [MD5 128/128 AltiVec 4x]... (64xOMP) DONE
Raw:	494592 c/s real, 8171 c/s virtual

$ GOMP_CPU_AFFINITY=0-63 ../run/john -test -form=md5crypt
Will run 64 OpenMP threads
Benchmarking: md5crypt, crypt(3) $1$ [MD5 128/128 AltiVec 4x]... (64xOMP) DONE
Raw:	547840 c/s real, 8557 c/s virtual
```

I'll investigate.

**Comment by solardiz** [comment]
```
Program received signal SIGILL, Illegal instruction.
s1 (out4=<optimized out>, out3=<optimized out>, out2=<optimized out>, out1=<optimized out>, a6=..., a5=..., a4=..., a3=..., 
    a2=..., a1=...) at nonstd.c:101
101		vxor(xD89697CC, xFAB77A3F, x2221EDF3);
```

```
(gdb) disass $pc-12,$pc+12
Dump of assembler code from 0x1001040c to 0x10010424:
   0x000000001001040c <DES_bs_crypt_25._omp_fn.0+4236>:	xxlxor  vs9,vs7,vs13
   0x0000000010010410 <DES_bs_crypt_25._omp_fn.0+4240>:	xxland  vs28,vs6,vs0
   0x0000000010010414 <DES_bs_crypt_25._omp_fn.0+4244>:	xxlor   vs29,vs8,vs9
=> 0x0000000010010418 <DES_bs_crypt_25._omp_fn.0+4248>:	xxleqv  vs13,vs13,vs28
   0x000000001001041c <DES_bs_crypt_25._omp_fn.0+4252>:	xxlxor  vs10,vs10,vs29
   0x0000000010010420 <DES_bs_crypt_25._omp_fn.0+4256>:	xxlxor  vs1,vs8,vs1
```

Also e.g. `raw-md5` crashes, while `raw-sha1` works.  For `raw-md5`:

```
Benchmarking: Raw-MD5 [MD5 128/128 AltiVec 4x]... 
Program received signal SIGILL, Illegal instruction.
source (source=<optimized out>, binary=0x10a82730) at rawMD5_fmt_plug.c:255
255		memcpy(b, binary, sizeof(b));
```

```
(gdb) disass $pc-20,$pc+12
Dump of assembler code from 0x102594e0 to 0x10259500:
   0x00000000102594e0 <source+32>:	std     r0,16(r1)
   0x00000000102594e4 <source+36>:	stdu    r1,-208(r1)
   0x00000000102594e8 <source+40>:	ld      r10,0(r4)
   0x00000000102594ec <source+44>:	ld      r11,8(r4)
   0x00000000102594f0 <source+48>:	addi    r3,r1,112
=> 0x00000000102594f4 <source+52>:	stq     r10,112(r1)
   0x00000000102594f8 <source+56>:	bl      0x10312a30 <md5_unreverse>
   0x00000000102594fc <source+60>:	nop
```

**Comment by solardiz** [comment]
'xxleqv' is apparently one of 3 extra SIMD bitwise instructions added with AltiVec2 in fairly recent ISA 2.07. Maybe this CPU doesn't support it yet.

`stq` is a very basic 64-bit mode instruction, which is not valid in 32-bit mode. And indeed it appears we're running in 32-bit mode:

```
msr            0x800000000004d032	9223372036855091250
```

As far as I could find, bit 0 means 0: 32-bit, 1: 64-bit mode (here it's at 0), and bit 63 means 0: BE, 1: LE (here it's at 1). So it seems we're in 32-bit LE?! Looks like our build is confused. Many register values have higher bits set, `pc` is in the 32-bit range, and the processor in 32-bit LE(?) mode, yet the code contains 64-bit mode only instructions.

**Comment by solardiz** [comment]
```
$ fgrep -A5 CFLAGSX Makefile | head -7
CFLAGSX = -c -m64 -g -O2 -I/usr/local/include -DARCH_LITTLE_ENDIAN=0   -Wall -Wdeclaration-after-statement -fomit-frame-pointer --param allow-store-data-races=0 -Wno-deprecated-declarations  -Wunused-but-set-variable -std=gnu89 -D_POSIX_SOURCE -D_GNU_SOURCE -D_XOPEN_SOURCE=600    -fopenmp  -pthread  $(RELEASE_BLD) $(CPPFLAGS)
# CFLAGS for use on the main john.c file only
CFLAGS_MAIN = -DAC_BUILT -DJOHN_ALTIVEC $(CFLAGSX)
CFLAGS = -DAC_BUILT  -maltivec -mvsx -mpower8-vector -DJOHN_ALTIVEC $(CFLAGSX)
ASFLAGS = -DAC_BUILT -m64  -DARCH_LITTLE_ENDIAN=0  -maltivec -mvsx -mpower8-vector -DJOHN_ALTIVEC -c -fopenmp   
LIBS = 
LDFLAGS = -g -m64  -L/usr/local/lib -L/usr/local/lib64 -L/usr/lib64 -L/lib64 $(LIBS) 
```

**Comment by solardiz** [comment]
```
$ head -4 /proc/cpuinfo 
processor	: 0
cpu		: POWER7 (architected), altivec supported
clock		: 3550.000000MHz
revision	: 2.1 (pvr 003f 0201)
```

I'll try rebuilding without `-mpower8-vector`.

**Comment by solardiz** [comment]
Trying to build with `-mpower8-vector` removed from the configure-generated Makefile fails with:

```
simd-intrinsics.c: In function 'SIMDSHA512body':
simd-intrinsics.c:2371:17: error: Builtin function __builtin_altivec_vaddudm requires the -mpower8-vector option
         tmp1[i] = vadd_epi64(h[i],    w[i][(x)&0xf]);       \
                 ^
simd-intrinsics.c:2598:2: note: in expansion of macro 'SHA512_STEP'
```

and many other errors like it. Jim, I think I'll return this task to you now so that you figure out how to build for the original AltiVec (without AltiVec2, which is POWER8+), while ideally supporting both (since apparently the latter has some helpful added instructions). BTW, I notice `configure` checks whether gcc accepts these options, but does it ever check whether the CPU we're building on supports those instruction sets? I guess not. I guess it should, if we do similar things on x86.

I also don't know what's up with the 32- vs. 64-bit mode weirdness. We do include `-m64` in there, so I'd expect a fully 64-bit build. Are some source files possibly built with different flags so that the `-m64` is not passed to there? Or do we invoke something that resets the CPU to 32-bit mode in `raw-md5` somehow? That's puzzling. If we're lucky, it'd turn out to be some POWER8 opcode having weird effect on POWER7 - then it's just one issue for us to deal with, as per the previous paragraph, and this weird issue would be gone. However, if it stays we'll have to investigate further.

**Comment by solardiz** [comment]
> On the bash history subject, here's what I do on ALL machines

With default readline settings (no `.inputrc` file), you can press Ctrl-R and then start typing `./con` for a similar effect. To see even earlier commands matching the pattern, simply press Ctrl-R again after you're already in this mode.

**Comment by solardiz** [comment]
> I notice configure checks whether gcc accepts these options, but does it ever check whether the CPU we're building on supports those instruction sets? I guess not. I guess it should, if we do similar things on x86.

And we should probably report AltiVec vs. AltiVec2 in configure output and in john output, much like we do e.g. for SSE2 vs. SSE4.1.

**Comment by solardiz** [comment]
Per Wikipedia, ISA 2.07 (POWER8) also adds AES, SHA-2 (including SHA-512), and CRC instructions. We could use those as well.

File: src/ppc64.h [hunk 1]
```diff
@@ -45,18 +45,40 @@
 #define DES_COPY			0
 #define DES_BS_ASM			0
 #define DES_BS				1
-#define DES_BS_VECTOR			0
+#if defined(JOHN_ALTIVEC)
+#define DES_BS_EXPAND			3
```
Comment (solardiz) [inline src/ppc64.h]: `DES_BS_EXPAND` must be 0 or 1, never 3. We use it as a Boolean.

File: src/ppc64.h [hunk 2]
```diff
@@ -45,18 +45,40 @@
...
+#if 1
+#define DES_BS_VECTOR			2
+#define DES_BS_ALGORITHM_NAME		"DES 128/128 AltiVec"
+#elif 0
+/* It is likely unreasonable to use S-box expressions requiring vsel when this
+ * operation is only available in one of the two instruction sets.
+ * So let's revert to less demanding S-box expressions. */
```
Comment (solardiz) [inline src/ppc64.h]: Looks like you copied this comment from a file of mine, but did not copy the accompanying `#undef` / `#define` (would need to change the value of `DES_BS`, but in your case you already set it to 1, so this wouldn't apply). Please be more careful. Comments that no longer reflect code are at best unhelpful.

Comment (solardiz) [reply src/ppc64.h]: Even more importantly, I have already tuned this AltiVec stuff on real POWER hardware. So please don't try to re-tune it on an emulator. Just leave it as-is, merging into those `.h` files only the jumbo-specific additions.

File: src/ppc64.h [hunk 3]
```diff
@@ -45,18 +45,40 @@
...
 
-#define MD5_ASM				0
+#define MD5_ASM			0
 #define MD5_X2				1
-#define MD5_IMM				0
+#if HAVE_ALTIVEC
+#define MD5_IMM			1
+#else
+#define MD5_IMM			0
+#endif
```
Comment (solardiz) [inline src/ppc64.h]: What version of `ppc64.h` (or what else?) did you start with? I have recently revised it in core to set `MD5_IMM` to 1 even for non-AltiVec, as that proved faster now. I mentioned this detail on a GitHub issue a couple of weeks ago.

File: src/configure.ac [hunk 4]
```diff
@@ -354,7 +399,7 @@ dnl AC_CHECK_SIZEOF([int *function()]
 # pa-risc.h
 # ppc32.h
 # ppc32alt.h (-maltivec)
-# ppc64.h (-m64)
+# ppc64.h (-m64 -maltivec)
 # ppc64alt.h (-maltivec -faltivec)
```
Comment (solardiz) [inline src/configure.ac]: Weird. Why are we listing `-maltivec` along with `ppc64.h` (not only `ppc64alt.h`), but also list `-faltivec` only along with the other? I guess for no specific reason?

**Comment by solardiz** [comment]
Jim, I just re-read your comment on the other issue about you merging stuff from `ppc64alt.h` into `ppc64.h`. This makes sense to me. You just need to merge from and into my latest versions of these files, and merge a more consistent (sub)set of the DES settings. There are a lot of them under `#if 0` - you should either merge all of them or not merge any of them. Thanks!

File: src/ppc64.h (hunk 1 above)
Comment (jfoug) [reply src/ppc64.h]: That was my mistake, taking it from ppc64alt.h

File: src/configure.ac (hunk 4 above)
Comment (jfoug) [reply src/configure.ac]: In jumbo land, ppc64.h was planned on being the only header needed.  but if there are 3 options (normal, Altivec1, Altivec2), then I may not be able to pack it all in one header.

Comment (solardiz) [reply src/configure.ac]: It should be possible to pack all 3 in one header.

**Comment by jfoug** [comment]
> BTW, I notice configure checks whether gcc accepts these options, but does it ever check whether the CPU we're building on supports those instruction sets? I guess not. I guess it should, if we do similar things on x86.

No, there is no probing of CPU capabilities (yet).  Only what the compiler can do.

Also, we have probing turned of in john (#define CPU_CKECK 0  or something like that).  So we are running a bit blind there, on both sides of the fence.

**Comment by jfoug** [comment]
I tried to rebase my branch, and I am pretty sure I screwed the pooch :(

I would hate to loose the verbiage of this PR, but I believe what it is based upon is going to do one of the circular mess pulls if it is accepted.   @magnumripper, any suggestions?   I can easily create a new branch and put the changes into that branch, created a PR from that, then back reference to this PR (for notes), if that would avoid the likely mess this one will produce.

**Comment by magnumripper** [comment]
I told you to do a `git pull --rebase origin bleeding-jumbo`. @kholia told you the same. But you didn't, right?

Try doing that now. Chances are things will bend back to a much better state.

**Comment by jfoug** [comment]
That was exactly what I did.

**Comment by magnumripper** [comment]
Yes but then you merged the PPC64 branch from upstream instead of force-pushing to it. Voila, a local merge.

Presuming your local branch is in the same state as here, try just doing a `git reset --hard HEAD^`. This should get rid of the local merge and get things back. If everything looks kosher, do a `git push -f` to here and nothing else.

Note that you must never EVER force-push to the main branch but you are free to do it to your topic branches / pull requests.

**Comment by jfoug** [comment]
Ok, I did a checkout of ppc64 (so local branch was same).

I did a git reset --hard HEAD^

It now show nothing in git status, and says I am 3 pushes behind origin/ppc64  (but I assume that is jfoug/ppc64 as the origin).

Here is the EXACT text seen:

```
$ git checkout ppc64
Switched to branch 'ppc64'
Your branch is up-to-date with 'origin/ppc64'.

$ git reset --hard HEAD^
HEAD is now at 4233c990c configure : Additional work on PPC64 port

$ git status
On branch ppc64
Your branch is behind 'origin/ppc64' by 3 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)
Untracked files:
  (use "git add <file>..." to include in what will be committed)
```

So at this time, I would do a   ``` git push -f ```   and things will be back to 'normal'  ?

**Comment by jfoug** [comment]
Ok, @magnumripper I think things are back to what they were.  I will simply work through the difference issues, when making any other changes.

**Comment by jfoug** [comment]
Whoever 'fixed' my circleCI stuff, THANK YOU!

**Comment by magnumripper** [comment]
> Whoever 'fixed' my circleCI stuff, THANK YOU!

😆 I have no idea if I did anything (don't think so)

So is this stuff ready for merge now?

**Comment by jfoug** [comment]
Yes, this is done.

**Comment by frank-dittrich** [comment]
@magnumripper 
> On the bash history subject, here's what I do on ALL machines I have to deal with
```
$ cat ~/.inputrc 
"\e[6~": history-search-forward
"\e[5~": history-search-backward
```

BTW: The side effect of such a change on my ubuntu 16.04 is that `<ctrl>-<cursor_left>` and `<ctrl>-<cursor_rigth>` don't work any longer (move one word left/right on the command line):

```
$ echo test test test;5D;5D
```
and
```
$ ;5C;5C;5Cecho test test test
```

**Pull request merged**